src/commands/generate.rs

```diff
@@ -13,20 +16,76 @@ pub struct Command {
     pub arguments: Vec<String>,
 }
 
-pub fn prepare_command(cmd: &Command) -> Result<Vec<&str>, anyhow::Error> {
-    let (lang, generator) = match cmd.generator.split_once("/") {
-        Some(v) => v,
-        None => anyhow::bail!("Generator should be of form <lang>/<tool>"),
-    };
+/// How we should invoke via `uv run` (or not).
+#[derive(Debug, Clone, Copy)]
+enum UseUv {
+    Always,
+    Auto,
+    Never,
+}
+
+impl UseUv {
+    /// Read from the `GEL_GENERATE_USE_UV` env var, defaulting to `auto`
+    fn from_env() -> Self {
+        match env::var("GEL_GENERATE_USE_UV")
+            .unwrap_or_default()
+            .to_lowercase()
+            .as_str()
+        {
+            "always" => UseUv::Always,
+            "never" => UseUv::Never,
+            _ => UseUv::Auto,
+        }
+    }
+}
+
+pub fn prepare_command(cmd: &Command) -> Result<Vec<OsString>, anyhow::Error> {
+    let (lang, generator) = cmd
+        .generator
+        .split_once('/')
+        .context("generator should be of form <lang>/<tool>")?;
+
+    let use_uv = UseUv::from_env();
 
     let commands = if lang == "py" {
-        if generator != "queries" {
-            anyhow::bail!("Unknown Python generator {}", generator)
+        let gen_name = "gel-generate-py";
+
+        let uv_invocation = || -> Result<Vec<OsString>, anyhow::Error> {
+            let uv = which("uv").context("`uv` not found on PATH")?;
+            Ok(vec![
+                uv.into_os_string(),
+                OsString::from("run"),
+                gen_name.into(),
+                generator.into(),
+            ])
+        };
+
+        match use_uv {
+            UseUv::Always => {
+                // always use uv
+                uv_invocation()?
+            }
+            UseUv::Never => {
+                // never use uv: must have the binary directly
+                let direct =
+                    which(gen_name).with_context(|| format!("`{}` not found on PATH", gen_name))?;
+                vec![direct.into_os_string(), generator.into()]
+            }
+            UseUv::Auto => {
+                // fallback logic: try direct, then uv
+                if let Ok(direct) = which(gen_name) {
+                    vec![direct.into_os_string(), generator.into()]
+                } else {
+                    // fall back to uv
+                    uv_invocation()?
+                }
+            }
         }
-        // XXX: or should we use `uv run`?
-        vec!["gel-py"]
     } else if lang == "js" {
         vec!["npx", "@gel/generate", generator]
+            .into_iter()
+            .map(OsString::from)
+            .collect()
```

Comment:
rust! I left it as `Vec<&str>` because I didn't want to have to do this annoyingly verbose coercion, but also knew that any change might end up requiring a type change...

(Though I guess it wouldn't if we put the non-resolved names in the Vec)